### PR TITLE
[contrib] Updating `runtime` and Dropping `UtilsRuntime`

### PR DESCRIPTION
**Pull request by ppenna**
Description
-------------

This PR updates the `runtime` crate.

Summary of Changes
------------------------

- Updated runtime crate and thus dropped `UtilsRuntime` from `inetstack` 
- Added Rust format check to build system
- Updated Rust toolchain to match latest nightly
- Fixed various warnings that popped up after updating the Rust toolchain

Further Considerations
--------------------------

- In order to drop `UtilsRuntime`, I introduced additional arguments for random number generator support on the constructors of the `Inetstack`, `UdpPeer` and `TcpPeer`.

### CODE REVIEW

File: src/protocols/arp/peer.rs [hunk 1]
```diff
@@ -66,6 +65,7 @@ use ::std::{
 pub struct ArpPeer<RT: NetworkRuntime> {
     rt: RT,
     cache: Rc<RefCell<ArpCache>>,
+    #[allow(unused)]
```
Comment (BrianZill) [inline src/protocols/arp/peer.rs:68]: Just out of curiosity, why did you have to add this?

File: src/protocols/icmpv4/peer.rs [hunk 2]
```diff
@@ -113,15 +117,18 @@ pub struct Icmpv4Peer<RT: SchedulerRuntime + UtilsRuntime + NetworkRuntime + Clo
...
 }
 
-impl<RT: SchedulerRuntime + UtilsRuntime + NetworkRuntime + Clone + 'static> Icmpv4Peer<RT> {
+impl<RT: SchedulerRuntime + NetworkRuntime + Clone + 'static> Icmpv4Peer<RT> {
     /// Creates a new peer for handling ICMP.
-    pub fn new(rt: RT, arp: ArpPeer<RT>) -> Icmpv4Peer<RT> {
+    pub fn new(rt: RT, arp: ArpPeer<RT>, rng_seed: [u8; 32]) -> Icmpv4Peer<RT> {
         let (tx, rx) = mpsc::unbounded();
         let requests = ReqQueue::new();
+        let rng: Rc<RefCell<SmallRng>> = Rc::new(RefCell::new(SmallRng::from_seed(rng_seed)));
```
Comment (BrianZill) [inline src/protocols/icmpv4/peer.rs:131]: It looks like we're now creating three separate random number generators (one each for ICMPv4, UDP, and TCP), all from the same seed.  Assuming they're deterministic, they will each spit out the same pseudo random numbers.  Why have three?  Couldn't we just create one up top and pass a pointer to it down to each place it is needed?  Instead of passing the seed down?  We could save some memory or use that memory budget on a better RNG (generally speaking, the more state it keeps around, the better the random number generator).

File: src/protocols/ipv4/datagram.rs [hunk 3]
```diff
@@ -73,7 +73,7 @@ pub struct Ipv4Header {
     /// Explicit Congestion Notification (2 bits).
     ecn: u8,
     /// Total length of the packet including header and data (16 bits).
-    total_length: u16,
+    _total_length: u16,
```
Comment (BrianZill) [inline src/protocols/ipv4/datagram.rs:76]: Why this change?

File: src/lib.rs [hunk 4]
```diff
@@ -87,20 +86,21 @@ pub mod protocols;
...
     rt: RT,
     ts_iters: usize,
 }
 
-impl<RT: SchedulerRuntime + UtilsRuntime + NetworkRuntime + Clone + 'static> InetStack<RT> {
+impl<RT: SchedulerRuntime + NetworkRuntime + Clone + 'static> InetStack<RT> {
     pub fn new(rt: RT) -> Result<Self, Fail> {
         let now: Instant = rt.now();
         let file_table: IoQueueTable = IoQueueTable::new();
+        let rng_seed: [u8; 32] = [0; 32];
```
[the remark this answers is not in the document]
Comment (ppenna) [reply src/lib.rs:101]: How about making this an argument of the constructor?

File: src/protocols/arp/peer.rs (hunk 1 above)
Comment (ppenna) [reply src/protocols/arp/peer.rs:68]: Because I updated the rust-toolchain and this was not showing as warning before.

The rust compiler thinks that the `background()` task is ever called and complains about it.

File: src/protocols/ipv4/datagram.rs (hunk 3 above)
Comment (ppenna) [reply src/protocols/ipv4/datagram.rs:76]: Because I updated the rust-toolchain and this was not showing as warning before.

The rust compiler thinks that the `total_length` is never read.

File: src/protocols/icmpv4/peer.rs (hunk 2 above)
Comment (ppenna) [reply src/protocols/icmpv4/peer.rs:131]: > Why have three?

I wish we have just one, and they way to go would be to wrap them around a `Rc<RefCell<SmallRng>>`.

Unfortunately, the `EphemeralPorts` which is used in both `UdpPeer` and `TcpPeer` use the `huffle<R>(&mut self, rng: &mut R)` function initially shuffle ports.

Once we wrap the pseudo-random number generator on the `Rc<RefCell<>>` we cannot comply with the signature of this function.

To get a single pseudo-random number generator, we should provide a better `EphemeralPorts`` abstraction, something that we've been already discussing.

File: src/lib.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (BrianZill) [reply src/lib.rs:101]: Yes, I suppose it could reasonably be considered a configuration parameter for the InetStack.

Of course, that just passes the buck over to the demikernel repo, where the CatnipLibOS constructor and the CatpowderLibOS constructor will need to pass it in.  I suppose we could always add this to the Config struct.  That would probably be the best solution.

File: src/protocols/arp/peer.rs (hunk 1 above)
Comment (BrianZill) [reply src/protocols/arp/peer.rs:68]: Can you add a short comment explaining this?

Comment (ppenna) [reply src/protocols/arp/peer.rs:68]: Sure.

File: src/protocols/ipv4/datagram.rs (hunk 3 above)
Comment (BrianZill) [reply src/protocols/ipv4/datagram.rs:76]: I see.  We trim the buffer during `parse()`, so we don't need to read `total_length` later to do that.  And we supply the payload length separately during serialization, so we don't read it there either.  There's an argument that `parse()` shouldn't be trimming the buffer and that `parse`'s caller should do that explicitly (in which case `total_length` would be read to do that).  Conversely, there's an argument that if nothing is reading the `total_length`, we should just remove it from the `Ipv4Header` struct (which I find unsatisfying as it is an important part of the on-wire representation).

If we leave it as is, I think I'd feel better with `#[allow(unused)]` and a comment vs. the underscore though.  Or maybe we should add a debug print of this field somewhere just to get the compiler to quit complaining (I've never liked those types of solutions to these problems though).

File: src/protocols/ipv4/datagram.rs [hunk 5]
```diff
@@ -105,13 +105,13 @@ impl Ipv4Header {
...
             ecn: 0,
-            total_length: IPV4_HEADER_MIN_SIZE,
+            _total_length: IPV4_HEADER_MIN_SIZE,
             identification: 0,
             flags: IPV4_CTRL_FLAG_DF,
             fragment_offset: 0,
             ttl: DEFAULT_IPV4_TTL,
             protocol,
-            header_checksum: 0,
+            _header_checksum: 0,
```
Comment (BrianZill) [inline src/protocols/ipv4/datagram.rs:114]: I have basically the same concerns about `header_checksum` that I did for `total_length`.  Although the argument for just removing it from the `Ipv4Header` struct is stronger here, as the on-wire representation is naturally checked upon parsing and written upon serialization.  So there's no need to have it in the `Ipv4Header` struct.

Maybe just remove it?

File: src/protocols/icmpv4/peer.rs (hunk 2 above)
Comment (BrianZill) [reply src/protocols/icmpv4/peer.rs:131]: There's got to be a reasonable way to do this in Rust.  Why doesn't `borrow_mut()` work on the `RefCell<>` holding the RNG?

Comment (ppenna) [reply src/protocols/icmpv4/peer.rs:131]: - To provide multiple ownership we either provide lifetime annotations across the whole stack OR we wrap the thing on a `Rc`
- To provide inner mutability, we should rely on `RefCell`
- `borrow_mut()` return the thing wrapped on a `RefMut` https://doc.rust-lang.org/std/cell/struct.RefCell.html#method.borrow_mut

File: src/protocols/ipv4/datagram.rs (hunk 3 above)
Comment (ppenna) [reply src/protocols/ipv4/datagram.rs:76]: I agree that dropping these fields of the structures is not a good idea and that we should annotate instead of using the prefix trick.

File: src/lib.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (ppenna) [reply src/lib.rs:101]: I changed it as we discussed.

File: src/protocols/icmpv4/peer.rs (hunk 2 above)
Comment (BrianZill) [reply src/protocols/icmpv4/peer.rs:133]: Yes, I don't understand why you said that couldn't work.

File: src/protocols/arp/peer.rs (hunk 1 above)
Comment (BrianZill) [reply src/protocols/arp/peer.rs]: Great, thanks!

File: src/protocols/ipv4/datagram.rs (hunk 3 above)
Comment (BrianZill) [reply src/protocols/ipv4/datagram.rs]: Okay, thanks!

File: src/lib.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (BrianZill) [reply src/lib.rs]: This is great, thanks!

File: src/protocols/icmpv4/peer.rs (hunk 2 above)
Comment (ppenna) [reply src/protocols/icmpv4/peer.rs:133]: No.

As I said, the shuffle function **requires** a mutable reference not a `RefMut<T>`

Comment (ppenna) [reply src/protocols/icmpv4/peer.rs:133]: @BrianZill here is how we do in Demikernel:

https://github.com/demikernel/demikernel/blob/559a8e1e13b0f260aa4ed050d1ad48074c381a63/src/catnip/runtime/utils.rs#L34-L37

Comment (BrianZill) [reply src/protocols/icmpv4/peer.rs:133]: Okay, I think we've wasted enough time trying to optimize something we want to eventually rewrite anyway.  There's no point to figuring out how to be more efficient around rng use when we're going to change how the rng is used by the ephemeral port allocator.
